### ISSUE TITLE
Bug: grad(jit(custom_linear_solve)) encounters ad_util.Zero in backwards pass

### ISSUE THREAD

**Issue by shoyer** [google/jax#1536]
I'm not quite sure what specifically is going wrong yet, but here's an example traceback from https://github.com/google/jax/pull/1535:

    jax/test_util.py:179: in check_grads
        _check_grads(f, args, order)
    jax/test_util.py:177: in _check_grads
        _check_grads(f_vjp, args, order - 1)
    jax/test_util.py:172: in _check_grads
        _check_vjp(f, partial(api.vjp, f), args)
    jax/test_util.py:148: in check_vjp
        cotangent_out = conj(vjpfun(conj(cotangent)))
    jax/api_util.py:66: in apply_flat_fun_nokwargs
        ans = fun(*args)
    jax/interpreters/ad.py:114: in vjp_
        _, arg_cts = backward_pass(jaxpr, consts, (), dummy_args, dummy_primals_and_cts)
    jax/interpreters/ad.py:178: in backward_pass
        eqn.params, subjaxpr, sub_consts, sub_freevar_vals, invals, cts_in)
    jax/interpreters/ad.py:458: in call_transpose
        out_flat = primitive.bind(fun, *all_args, **params)
    jax/core.py:591: in call_bind
        outs = primitive.impl(f, *args, **params)
    jax/interpreters/xla.py:418: in _xla_call_impl
        compiled_fun = _xla_callable(fun, device, backend, *map(abstractify, args))
    jax/linear_util.py:217: in cached_fun
        ans, f_prev = cached_fun_body(f, args)
    jax/linear_util.py:214: in cached_fun_body
        return call(f, *args), f
    jax/interpreters/xla.py:434: in _xla_callable
        jaxpr, (pvals, consts, env) = pe.trace_to_subjaxpr(fun, master, False).call_wrapped(pvals)
    jax/linear_util.py:165: in call_wrapped
        ans = self.f(*args, **dict(self.params, **kwargs))
    jax/interpreters/ad.py:178: in backward_pass
        eqn.params, subjaxpr, sub_consts, sub_freevar_vals, invals, cts_in)
    jax/interpreters/ad.py:458: in call_transpose
        out_flat = primitive.bind(fun, *all_args, **params)
    jax/core.py:594: in call_bind
        outs = map(full_lower, top_trace.process_call(primitive, f, tracers, params))
    jax/interpreters/partial_eval.py:111: in process_call
        out_flat = call_primitive.bind(fun, *in_consts, **params)
    jax/core.py:591: in call_bind
        outs = primitive.impl(f, *args, **params)
    jax/interpreters/xla.py:418: in _xla_call_impl
        compiled_fun = _xla_callable(fun, device, backend, *map(abstractify, args))
    jax/linear_util.py:217: in cached_fun
        ans, f_prev = cached_fun_body(f, args)
    jax/linear_util.py:214: in cached_fun_body
        return call(f, *args), f
    jax/interpreters/xla.py:434: in _xla_callable
        jaxpr, (pvals, consts, env) = pe.trace_to_subjaxpr(fun, master, False).call_wrapped(pvals)
    jax/linear_util.py:165: in call_wrapped
        ans = self.f(*args, **dict(self.params, **kwargs))
    jax/interpreters/ad.py:181: in backward_pass
        cts_out = get_primitive_transpose(eqn.primitive)(cts_in, *invals, **eqn.params)
    jax/lax_linalg.py:358: in triangular_solve_transpose_rule
        not transpose_a, conjugate_a, unit_diagonal)
    jax/lax_linalg.py:80: in triangular_solve
        conjugate_a=conjugate_a, unit_diagonal=unit_diagonal)
    jax/core.py:155: in bind
        out_tracer = top_trace.process_primitive(self, tracers, kwargs)
    jax/interpreters/partial_eval.py:91: in process_primitive
        tracers = map(self.instantiate_const, tracers)
    jax/util.py:39: in safe_map
        return list(map(f, *args))
    jax/interpreters/partial_eval.py:80: in instantiate_const
        return self.new_instantiated_const(const)
    jax/interpreters/partial_eval.py:66: in new_instantiated_const
        return JaxprTracer(self, PartialVal((get_aval(val), unit)), ConstVar(val))
    jax/core.py:538: in get_aval
        return concrete_aval(x)
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    x = Zero

        def concrete_aval(x):
          try:
            return pytype_aval_mappings[type(x)](x)
          except KeyError:
    >       raise TypeError("{} is not a valid Jax type".format(type(x)))
    E       TypeError: <class 'jax.ad_util.Zero'> is not a valid Jax type

**Comment by jekbradbury** [comment]
Transpose rules need to explicitly deal with symbolic zeros as input cotangents, since ordinary JAX code can't generally handle them. It looks like `triangular_solve_transpose_rule` doesn't?

**Comment by shoyer** [comment]
> The primitives in lax.py tend to have [explicit handling](https://github.com/google/jax/blob/master/jax/lax/lax.py#L1735) of `ad_util.zero` but It looks like `triangular_solve_transpose_rule` (and the other linalg primitives) doesn't?

Hmm, maybe all I need to do to fix this is handle `ad_util.zero` explicitly in the transpose rule? I wasn't sure if transpose rules needed to handle symbolic zeros.

I think the main reason why you don't need symbolic zeros in `lax_linalg.py` right now is that I convinced myself that they are never needed in JVP rules and removed them:
https://github.com/google/jax/pull/1490

**Comment by shoyer** [comment]
Any suggestions for how I could reproduce this with triangular solve directly? I tried this in `linalg_tests.py`, but it did not suffice: 
```
  def testTriangularSolveGradZeros(self):
    a_shape = (3, 3)
    b_shape = (1, 3)
    dtype = onp.float32
    rng = jtu.rand_default()
    a = np.triu(rng(a_shape, dtype) + onp.eye(a_shape[-1], dtype=dtype))
    b = rng(b_shape, dtype)
    f = lax_linalg.triangular_solve
    jtu.check_grads(lambda x: f(a, x), (b,), 2, rtol=2e-4, eps=1e-3)
    jtu.check_grads(lambda x: f(x, b), (a,), 2, rtol=2e-4, eps=1e-3)
```

**Comment by shoyer** [comment]
OK, this was easy enough to fix in #1537 

I do wonder if we might be able to rule out this class of bugs entirely by not calling primitive transpose rules if `cotangent` is a symbolic zero, like we currently seem to do for JVPs.
